pennylane/clusters.py

```diff
@@ -0,0 +1,476 @@
+        # Generate QNodes from all pairs of observables, measurements, and devices.
+        if not isinstance(obs, Observable):
+            raise ValueError("Could not create QNodes. Some or all observables are not valid.")
+
+        wires = list(range(dev.num_wires))
+
+        def circuit(
+            params, _obs=obs, _m=m, _wires=wires, **kwargs
+        ):  # pylint: disable=dangerous-default-value, function-redefined
+            template(params, wires=_wires, **kwargs)
+            return MEASURE_MAP[_m](_obs)
+
+        qnode = QNode(circuit, dev, interface=interface, diff_method=diff_method)
+        qnodes.append(qnode)
+
+    return qnodes
+
+
+def apply(func, qnode_cluster):
+    """Lazily apply a function to the constituent QNodes of a :class:`QNodeCluster`.
+
+    Args:
+        func (callable): A function to be applied to the QNodeCluster results.
+            This function must be supported by the corresponding QNodeCluster
+            interface; i.e., a ``torch`` QNodeCluster can only be acted on functions
+            that accept ``torch.tensor`` objects.
+        qnode_cluster (QNodeCluster): a QNode cluster.
+
+    **Example:**
+
+    We can create a QNodeCluster using :func:`~.map`:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(1)]
+    >>> qnodes = qml.map(qml.templates.StronglyEntanglingLayers, obs_list, dev, interface="torch")
+
+    As we are using the ``'torch'`` interface, we now apply ``torch.sum``
+    to the QNodeCluster:
+
+    >>> cost = qml.apply(torch.sum, qnodes)
+
+    This is a lazy composition --- no QNode evaluation has yet occured. Evaluation
+    only occurs when the returned function ``cost`` is evaluated:
+
+    >>> x = qml.init.strong_ent_layers_normal(3, 2)
+    >>> cost(x)
+    tensor(0.9092, dtype=torch.float64, grad_fn=<SumBackward0>)
+    """
+    return lambda params, **kwargs: func(qnode_cluster(params, **kwargs))
+
+
+def sum(x):
+    """Lazily sum the constituent QNodes of a :class:`QNodeCluster`.
+
+    Args:
+        x (QNodeCluster): a QNode cluster of independent QNodes.
+
+    .. seealso:: :func:`~.apply`, :func:`~.dot`
+
+    **Example:**
+
+    We can create a QNodeCluster using :func:`~.map`:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(1)]
+    >>> qnodes = qml.map(qml.templates.StronglyEntanglingLayers, obs_list, dev, interface="torch")
+
+    For the cost function, we now sum the results of all QNodes in the cluster:
+
+    >>> cost = qml.sum(qnodes)
+
+    This is a lazy summation --- no QNode evaluation has yet occured. Evaluation
+    only occurs when the returned function ``cost`` is evaluated:
+
+    >>> x = qml.init.strong_ent_layers_normal(3, 2)
+    >>> cost(x)
+    tensor(0.9092, dtype=torch.float64, grad_fn=<SumBackward0>)
+    """
+    if hasattr(x, "interface") and x.interface is not None:
```

Comment:
It might be worth adding input validation for ``x`` such as:
```
if not isinstance(x, qml.QNodeCluster):
    raise ValueError("Unknown qnode cluster passed {}".format(x))
```
or something along those lines.

It might be even worth considering accepting a list of ``QNodes`` that is converted into a ``QNodeCluster`` by default.

Reply:
Oh that is a good idea

Reply:
I thought about it a bit more, and maybe we can think of adding it in a separate PR?

The reason being is that we should encourage usage of `map` to create the QNode cluster for use with `apply`, as opposed to passing a list of QNodes to `apply`. This is because the former enforces that all QNodes have the same signature, whereas with the latter, it might be easier to accidentally create a list of QNodes with different signatures, leading to weird exceptions.

Although I'm still on the fence, if you have any additional thoughts?

Reply:
(Numbered the two points in my original comment such that we can refer to each explicitly)
1. What would your thoughts be about validation?
2. I agree this could be something coming up later as well!

Reply:
If validation was simple, I would be very much for it. However, I can't think of a way to do the validation that doesn't require complicated usage of `inspect`. Even then, if the signatures match, it's hard to know if the QNodes were designed so that they take the same size arguments...

Reply:
My bad, I was more thinking of an  ``isinstance(x, QNodeCluster)`` validation, through calling ``apply`` there is going to be validation taking place anyways (assuming that an object that has an ``interface`` attribute is passed, else an ``AttributeError`` is going to be raised). This latter case might be unwanted, as the user might want another error message (though might not be crucial).